### PR TITLE
Move schedulers to training from modules.

### PR DESCRIPTION
**Pull request by krammnic**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [X] fix a bug
- [X] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
#1587 

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [X] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [X] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [X] run unit tests via `pytest tests`
- [X] run recipe tests via `pytest tests -m integration_test`
- [X] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [X] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by krammnic** [comment]
@joecummings Require review

**Comment by krammnic** [comment]
Should include backward capability, no merge right now!

**Comment by krammnic** [comment]
@ebsmothers Fixed backward capability

**Review by ebsmothers — COMMENTED** [review]
Can you also update the usage [here](https://github.com/pytorch/torchtune/blob/c5b738681ef0c2541344aff1063e5f85720b3c1f/docs/source/tutorials/memory_optimizations.rst?plain=1#L131) and [here](https://github.com/pytorch/torchtune/blob/c5b738681ef0c2541344aff1063e5f85720b3c1f/recipes/configs/qwen2/knowledge_distillation_single_device.yaml#L70)?

**Review by ebsmothers — COMMENTED** [review]
Oh actually I think the new version under `torchtune/training/lr_schedulers.py` is missing from the PR. Also separately if you are running into issues with your local pre-commit not lining up with the CI in lint job, you can mimic [these steps](https://github.com/pytorch/torchtune/blob/main/.github/workflows/lint.yaml#L32-L40) in our CI job more directly on your local machine to try to get the same setup (though I'd think that the usual flow in our CONTRIBUTING.md around `pre-commit install` should be sufficient)

**Comment by krammnic** [comment]
Yeah, will do some fix soon. I have broken something with my commits

**Comment by krammnic** [comment]
> Oh actually I think the new version under `torchtune/training/lr_schedulers.py` is missing from the PR.
> 
> Also separately if you are running into issues with your local pre-commit not lining up with the CI in lint job, you can mimic [these steps](https://github.com/pytorch/torchtune/blob/main/.github/workflows/lint.yaml#L32-L40) in our CI job more directly on your local machine to try to get the same setup (though I'd think that the usual flow in our CONTRIBUTING.md around `pre-commit install` should be sufficient)

Fixed. Yeah, probably I will mimic CI using instructions that you provided, because the way that is provided in CONTRIBUTING.md is pretty inconvenient for me.

**Comment by ebsmothers** [comment]
> > Oh actually I think the new version under `torchtune/training/lr_schedulers.py` is missing from the PR.
> > Also separately if you are running into issues with your local pre-commit not lining up with the CI in lint job, you can mimic [these steps](https://github.com/pytorch/torchtune/blob/main/.github/workflows/lint.yaml?rgh-link-date=2024-10-11T11%3A37%3A22Z#L32-L40) in our CI job more directly on your local machine to try to get the same setup (though I'd think that the usual flow in our CONTRIBUTING.md around `pre-commit install` should be sufficient)
> 
> Fixed. Yeah, probably I will mimic CI using instructions that you provided, because the way that is provided in CONTRIBUTING.md is pretty inconvenient for me.

Sounds good. By the way if CONTRIBUTING.md is not currently clear or giving the right setup you need, we should update it. We set it up a while back and many devs don't use it much day-to-day. But as someone onboarding to the project recently if there are certain areas where you're feeling friction let us know and we can hopefully make the process smoother.

**Comment by krammnic** [comment]
@ebsmothers Which label to put for schedulers section in docs? Will fix lint now and probably except this point it will be ready to merge.

File: docs/source/api_ref_training.rst
Comment (ebsmothers) [reply docs/source/api_ref_training.rst:77]: @krammnic in response to your question
```suggestion
.. _lr_scheduler_label:
```

**Comment by ebsmothers** [comment]
> @ebsmothers Which label to put for schedulers section in docs? Will fix lint now and probably except this point it will be ready to merge.

Updated my inline comment [here](https://github.com/pytorch/torchtune/pull/1801#discussion_r1796976806) with the suggestion

**Comment by krammnic** [comment]
Tag is fixed.

**Comment by krammnic** [comment]
Last lint fix. Should be fine right now

File: docs/source/api_ref_training.rst
Comment (RdoubleA) [inline docs/source/api_ref_training.rst:90]: this should swap with metric_logging_label above, and should have an underscore before: `.. _lr_scheduler_label:`. otherwise it gets interpreted as a comment (sphinx, I know)

File: docs/source/api_ref_training.rst
Comment (RdoubleA) [inline docs/source/api_ref_training.rst:80]: the underline needs to match the length of the "Schedulers" title

File: docs/source/api_ref_training.rst
Comment (krammnic) [reply docs/source/api_ref_training.rst]: Fixed

File: docs/source/api_ref_training.rst
Comment (krammnic) [reply docs/source/api_ref_training.rst]: Fixed

**Comment by krammnic** [comment]
Finnaly! Can be merged...

**Comment by krammnic** [comment]
@RdoubleA As you already reviewed this maybe we can merge?

File: docs/source/api_ref_training.rst
Comment (ebsmothers) [inline docs/source/api_ref_training.rst:90]: Is this right? Seems to me like lr scheduler label and metric logging label need to be swapped

Comment (krammnic) [reply docs/source/api_ref_training.rst:90]: Are you speaking about 2 sections?

**Comment by joecummings** [comment]
@RdoubleA and/or @ebsmothers can one of you sign off on this?

**Comment by ebsmothers** [comment]
> @RdoubleA and/or @ebsmothers can one of you sign off on this?

@joecummings I have one more comment on the file api_ref_training.rst. After that I am good to merge this

File: docs/source/api_ref_training.rst
Comment (krammnic) [reply docs/source/api_ref_training.rst]: I seen now what was your idea, fixed!

**Review by ebsmothers — APPROVED** [review]
Thanks for making this change! Once CI is green I think we're good to merge here

**Pull request merged**
